File: analyze_excel.py
#!/usr/bin/env python3
import openpyxl
from openpyxl import load_workbook
import json

def analyze_payroll_excel(filepath):
    """Analyze the Excel payroll sheet and extract formulas"""
    wb = load_workbook(filepath, data_only=False)
    sheet = wb.active
    
    analysis = {
        "sheet_name": sheet.title,
        "columns": [],
        "formulas": {},
        "attendance_logic": {},
        "calculation_methods": {}
    }
    
    # Get column headers
    headers = {}
    for col in range(1, sheet.max_column + 1):
        cell = sheet.cell(row=1, column=col)
        if cell.value:
            headers[col] = cell.value
            analysis["columns"].append(cell.value)
    
    # Analyze formulas in the sheet
    for row in sheet.iter_rows(min_row=2, max_row=min(sheet.max_row, 10)):
        for cell in row:
            if cell.value and isinstance(cell.value, str) and cell.value.startswith('='):
                col_name = headers.get(cell.column, f"Column {cell.column}")
                if col_name not in analysis["formulas"]:
                    analysis["formulas"][col_name] = []
                analysis["formulas"][col_name].append({
                    "cell": f"{cell.column_letter}{cell.row}",
                    "formula": cell.value
                })
    
    # Look for specific attendance-related formulas
    attendance_columns = []
    for col, header in headers.items():
        if any(keyword in str(header).lower() for keyword in ['attendance', 'present', 'absent', 'working', 'days']):
            attendance_columns.append((col, header))
    
    # Extract attendance calculation logic
    for col, header in attendance_columns:
        formulas = []
        for row in range(2, min(sheet.max_row + 1, 10)):
            cell = sheet.cell(row=row, column=col)
            if cell.value and isinstance(cell.value, str) and cell.value.startswith('='):
                formulas.append(cell.value)
        if formulas:
            analysis["attendance_logic"][header] = formulas
    
    # Look for salary calculation patterns
    salary_columns = []
    for col, header in headers.items():
        if any(keyword in str(header).lower() for keyword in ['salary', 'basic', 'hra', 'allowance', 'gross', 'net', 'deduction']):
            salary_columns.append((col, header))
    
    # Extract salary calculation methods
    for col, header in salary_columns:
        formulas = []
        for row in range(2, min(sheet.max_row + 1, 10)):
            cell = sheet.cell(row=row, column=col)
            if cell.value and isinstance(cell.value, str) and cell.value.startswith('='):
                formulas.append(cell.value)
        if formulas:
            analysis["calculation_methods"][header] = formulas[0] if formulas else None
    
    # Print analysis
    print("=== EXCEL PAYROLL SHEET ANALYSIS ===\n")
    print(f"Sheet Name: {analysis['sheet_name']}")
    print(f"Total Columns: {len(analysis['columns'])}")
    print("\nColumn Headers:")
    for i, col in enumerate(analysis['columns'], 1):
        print(f"  {i}. {col}")
    
    print("\n=== ATTENDANCE FORMULAS ===")
    if analysis["attendance_logic"]:
        for col, formulas in analysis["attendance_logic"].items():
            print(f"\n{col}:")
            for formula in formulas[:3]:  # Show first 3 examples
                print(f"  {formula}")
    else:
        print("No attendance formulas found")
    
    print("\n=== SALARY CALCULATION FORMULAS ===")
    if analysis["calculation_methods"]:
        for col, formula in analysis["calculation_methods"].items():
            print(f"\n{col}:")
            print(f"  {formula}")
    else:
        print("No salary calculation formulas found")
    
    # Save detailed analysis
    with open('/home/bhara/payroll/excel_analysis.json', 'w') as f:
        json.dump(analysis, f, indent=2)
    
    return analysis

# Analyze the file
try:
    analyze_payroll_excel('/mnt/c/Users/bhara/Downloads/Payroll_June_25.xlsx')
except Exception as e:
    print(f"Error analyzing Excel file: {e}")
    import traceback
    traceback.print_exc()

File: analyze_excel.js
const XLSX = require('xlsx');
const fs = require('fs');

function analyzePayrollExcel(filepath) {
    console.log('=== ANALYZING PAYROLL EXCEL SHEET ===\n');
    
    try {
        // Read the workbook
        const workbook = XLSX.readFile(filepath);
        const sheetName = workbook.SheetNames[0];
        const worksheet = workbook.Sheets[sheetName];
        
        console.log(`Sheet Name: ${sheetName}`);
        
        // Get the range of the sheet
        const range = XLSX.utils.decode_range(worksheet['!ref']);
        
        // Extract headers
        const headers = [];
        for (let C = range.s.c; C <= range.e.c; ++C) {
            const cellAddress = XLSX.utils.encode_cell({r: 0, c: C});
            const cell = worksheet[cellAddress];
            headers.push(cell ? cell.v : `Column ${C + 1}`);
        }
        
        console.log(`\nColumn Headers (${headers.length} columns):`);
        headers.forEach((header, index) => {
            console.log(`  ${index + 1}. ${header}`);
        });
        
        // Look for formulas
        console.log('\n=== FORMULA ANALYSIS ===\n');
        
        const formulasByColumn = {};
        let totalFormulas = 0;
        
        for (let R = range.s.r; R <= Math.min(range.e.r, 10); ++R) {
            for (let C = range.s.c; C <= range.e.c; ++C) {
                const cellAddress = XLSX.utils.encode_cell({r: R, c: C});
                const cell = worksheet[cellAddress];
                
                if (cell && cell.f) {
                    const columnName = headers[C] || `Column ${C + 1}`;
                    if (!formulasByColumn[columnName]) {
                        formulasByColumn[columnName] = [];
                    }
                    formulasByColumn[columnName].push({
                        cell: cellAddress,
                        formula: cell.f,
                        value: cell.v
                    });
                    totalFormulas++;
                }
            }
        }
        
        console.log(`Total formulas found: ${totalFormulas}`);
        
        // Analyze attendance-related formulas
        console.log('\n=== ATTENDANCE-RELATED FORMULAS ===');
        const attendanceKeywords = ['attendance', 'present', 'absent', 'working', 'days', 'p', 'a', 'l'];
        
        Object.entries(formulasByColumn).forEach(([column, formulas]) => {
            if (attendanceKeywords.some(keyword => column.toLowerCase().includes(keyword))) {
                console.log(`\n${column}:`);
                formulas.slice(0, 3).forEach(({cell, formula}) => {
                    console.log(`  Cell ${cell}: ${formula}`);
                });
            }
        });
        
        // Analyze salary calculation formulas
        console.log('\n=== SALARY CALCULATION FORMULAS ===');
        const salaryKeywords = ['salary', 'basic', 'hra', 'allowance', 'gross', 'net', 'deduction', 'pf', 'esic', 'pt'];
        
        Object.entries(formulasByColumn).forEach(([column, formulas]) => {
            if (salaryKeywords.some(keyword => column.toLowerCase().includes(keyword))) {
                console.log(`\n${column}:`);
                if (formulas.length > 0) {
                    console.log(`  Example: ${formulas[0].formula}`);
                }
            }
        });
        
        // Extract specific attendance counting logic
        console.log('\n=== ATTENDANCE COUNTING LOGIC ===');
        
        // Look for COUNTIF formulas
        const countifFormulas = [];
        Object.entries(formulasByColumn).forEach(([column, formulas]) => {
            formulas.forEach(({formula, cell}) => {
                if (formula.includes('COUNTIF')) {
                    countifFormulas.push({column, formula, cell});
                }
            });
        });
        
        if (countifFormulas.length > 0) {
            console.log('\nCOUNTIF formulas found:');
            countifFormulas.forEach(({column, formula, cell}) => {
                console.log(`  ${column} (${cell}): ${formula}`);
            });
        }
        
        // Look for specific patterns in formulas
        console.log('\n=== FORMULA PATTERNS ===');
        
        // Check for daily rate calculations
        const dailyRateFormulas = [];
        Object.entries(formulasByColumn).forEach(([column, formulas]) => {
            formulas.forEach(({formula}) => {
                if (formula.includes('/30') || formula.includes('/31') || formula.includes('/28')) {
                    dailyRateFormulas.push({column, formula});
                }
            });
        });
        
        if (dailyRateFormulas.length > 0) {
            console.log('\nDaily rate calculations:');
            dailyRateFormulas.slice(0, 3).forEach(({column, formula}) => {
                console.log(`  ${column}: ${formula}`);
            });
        }
        
        // Save detailed analysis
        const analysis = {
            sheetName,
            headers,
            formulasByColumn,
            attendanceFormulas: countifFormulas,
            dailyRateCalculations: dailyRateFormulas
        };
        
        fs.writeFileSync('/home/bhara/payroll/excel_analysis.json', JSON.stringify(analysis, null, 2));
        console.log('\n✅ Detailed analysis saved to excel_analysis.json');
        
    } catch (error) {
        console.error('Error analyzing Excel file:', error);
    }
}

// Run the analysis
analyzePayrollExcel('/mnt/c/Users/bhara/Downloads/Payroll_June_25.xlsx');